tests/core/test_expectation_suite.py

```diff
@@ -362,6 +362,210 @@ def test_save_success(self):
+        context = Mock(spec=AbstractDataContext)
+        set_context(project=context)
+        context.expectations_store.has_key.return_value = True
+        old_expectation_configuration = expectation.configuration
+        suite = ExpectationSuite(
+            expectation_suite_name=self.expectation_suite_name,
+            expectations=[old_expectation_configuration],
+        )
+        store_key = context.expectations_store.get_key.return_value
+
+        # todo: this is the correct pattern, but depends on PR 9040
+        # expectation = suite.expectations[0]
+        # expectation.column = "b"
+        expectation = ExpectColumnValuesToBeInSet(
+            column="b",
+            value_set=[1, 2, 3],
+            result_format="BASIC",
+        )
+
+        suite.update_by_identity(
+            identity=id(old_expectation_configuration), expectation=expectation
+        )
+
+        # expect that the data context is kept in sync
+        context.expectations_store.add_or_update.assert_called_once_with(
+            key=store_key, value=suite
+        )
+
+    @pytest.mark.unit
+    def test_update_by_identity_doesnt_persist_unsaved_suite(self, expectation):
+        context = Mock(spec=AbstractDataContext)
+        set_context(project=context)
+        context.expectations_store.has_key.return_value = False
+        old_expectation_configuration = expectation.configuration
+        suite = ExpectationSuite(
+            expectation_suite_name=self.expectation_suite_name,
+            expectations=[old_expectation_configuration],
+        )
+
+        # todo: this is the correct pattern, but depends on PR 9040
+        # expectation = suite.expectations[0]
+        # expectation.column = "b"
+        expectation = ExpectColumnValuesToBeInSet(
+            column="b",
+            value_set=[1, 2, 3],
+            result_format="BASIC",
+        )
+
+        suite.update_by_identity(
+            identity=id(old_expectation_configuration), expectation=expectation
+        )
+        context.expectations_store.add_or_update.assert_not_called()
+
+    @pytest.mark.unit
+    def test_update_by_identity_fails_on_missing_expectation(self, expectation):
+        context = Mock(spec=AbstractDataContext)
+        set_context(project=context)
+        context.expectations_store.has_key.return_value = True
+        old_expectation_configuration = expectation.configuration
+        suite = ExpectationSuite(
+            expectation_suite_name=self.expectation_suite_name,
+            expectations=[],
+        )
+
+        # todo: this is the correct pattern, but depends on PR 9040
+        # expectation = suite.expectations[0]
+        # expectation.column = "b"
+        expectation = ExpectColumnValuesToBeInSet(
+            column="b",
+            value_set=[1, 2, 3],
+            result_format="BASIC",
+        )
+
+        with pytest.raises(
+            KeyError, match="Cannot update Expectation because it was not found."
```

Comment:
Just flagging here for anyone interested that we discussed raising a more descriptive error.